File: September 2022/Day 14/C++ Code/1457. Pseudo-Palindromic Paths in a Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    int pseudoPalindromicPaths (TreeNode* root) {
        vector<int> fre(10,0);
        return solve(root,fre);
    }
    
    int solve(TreeNode* root,vector<int>& fre) {
        if(root==NULL) return 0;
        
        if(root->left==NULL && root-> right==NULL) {
            fre[root->val]++;
            bool isValid = isPalindrome(fre);
            fre[root->val]--;
            if(isValid) return 1;
            return 0;
        }
        
        fre[root->val]++;
        int lft = solve(root->left,fre);
        int rt = solve(root->right,fre);
        fre[root->val]--;
        return lft+rt;
    }
    
    bool isPalindrome(vector<int>& a) {
        int odd=0;
        for(int i=0;i<a.size();i++) {
            if(a[i]%2==1) odd++;
        }
        if(odd>1) return false;
        return true;
    }
    
};

File: September 2022/Day 14/JAVA Code/1457. Pseudo-Palindromic Paths in a Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int pseudoPalindromicPaths (TreeNode root) {
        return solve(root,new int[10]);
    }
    
    public static int solve(TreeNode root,int[] fre) {
        if(root==null) return 0;
        if(root.left==null && root.right==null) {
            fre[root.val]++;
            if(isPalindrome(fre)) {
                fre[root.val]--;
                return 1;
            }
            fre[root.val]--;
            return 0;
        }
        fre[root.val]++;
        int v1 = solve(root.left,fre);
        int v2 = solve(root.right,fre);
        fre[root.val]--;
        return v1+v2;
    }
    
    public static boolean isPalindrome(int[] a) {
        int odd=0;
        for(int v : a) {
            if(v%2==1) odd++;
        }
        if(odd>1) return false;
        return true;
    }
}
